### ISSUE TITLE
Should warn or error on `isinstance` checks that can never be `True`

### ISSUE THREAD

**Issue by ottumm** [microsoft/pyright#4539]
The following does not generate any errors, even though `isinstance(a, str)` can never be `True`.

```python
a = 5

assert isinstance(a, str)

a.startswith("foo")
```
([pyright playground](https://pyright-playground.decorator-factory.su/?gzip=H4sIAHo71GMC_0tUsFUw5eJKLC5OLSpRyCzOzCsuScxLTtVI1FEoLinSBErpAUWKSorLM0syNJTS8vOVNLkANvf43DYAAAA%3D))

It would be wonderful if Pyright could flag the `isinstance` call with a warning (or error in strict mode), because in effect it hides the following errors.

This came up for me when I was refactoring some code and changed a return type of a function. There were assertions similar to the above in the codebase that were doing type narrowing but that once I changed the return type were no longer valid (type narrowing would have been impossible). What could have been a static-time error was instead a runtime error.

**Comment by ottumm** [comment]
I think this would be analogous to (or part of) `reportUnnecessaryComparison` and `reportUnnecessaryContains`.

**Comment by erictraut** [comment]
The behavior is intended. 

Pyright already supports a check called `reportUnnecessaryIsInstance`. However, it will not generate a diagnostic in your case for two reasons. First, this check is intentionally disabled for expressions used within an `assert` statement. Asserts are often used in cases where static types are inaccurate. Second, it is possible for a class to subclass from both int and str, so it would be incorrect to conclude that `isinstance` is unnecessary in this case. The `reportUnnecessaryIsInstance` is conservative and flags only cases where the `isinstance` is truly unnecessary.

Here is an example:
```python
# pyright: strict

a = 5

if isinstance(a, int):
    pass
```

**Comment by ottumm** [comment]
Ahhh, once again I've forgotten about multiple inheritance, and once again it has bitten me. Thanks for the explanation!

**Comment by ottumm** [comment]
Hmmm, although, in the case I included, Pyright can be absolutely sure that the concrete type of `a` is `int`, right? And so `isinstance(a, str)` is provably false. So it does seem like `reportUnnecessaryIsInstance` should catch this case.

And it doesn't seem like it's just because of the `assert` that it's letting it through, because this also type-checks cleanly:

```python
a = 5

if not isinstance(a, str):
    assert(False)

a.startswith("foo")
```

**Comment by erictraut** [comment]
That's not how type narrowing works.

Pyright is working as designed here.

**Comment by ottumm** [comment]
I can see we're at an impasse. I get that it's working as designed, but I would like to point out that I didn't open a bug, I opened a feature request. So I'm asking if the design can change.

Obviously the sample code I gave is a gross over-simplification of what I actually ran into in my codebase. Perhaps I simplified it too much, and perhaps I do in fact not understand how type narrowing works (ouch, by the way). But what I ran into was really a problem that type checking really could have caught. I'm quite sure of that.

**Comment by erictraut** [comment]
Can you provide a more realistic code sample — one that demonstrates the type of bug that you hit in your code?

**Comment by ottumm** [comment]
Yes, thank you, I'll put one together.

**Comment by ottumm** [comment]
Ok, here's an example that lines up very closely with what I experienced in the codebase I'm working in. Still highly simplified but conceptually the same:

```python
from collections.abc import Sequence


class FloatPolygon:
    def __init__(self, points: Sequence[tuple[float, float]]):
        self.points = points

    def first_float_point(self) -> tuple[float, float]:
        return self.points[0]


class IntPolygon:
    def __init__(self, points: Sequence[tuple[int, int]]):
        self.points = points

    def first_int_point(self) -> tuple[int, int]:
        return self.points[0]


class BothPolygon:
    def __init__(self, int_poly: IntPolygon, float_poly: FloatPolygon):
        self.int_poly = int_poly
        self.float_poly = float_poly


def old_polygon_function() -> FloatPolygon | IntPolygon:
    return FloatPolygon([(1.2, 1.5), (0, 0.5), (1.5, 3.4)])


def new_polygon_function() -> BothPolygon:
    return BothPolygon(
        IntPolygon([(0, 0), (1, 0), (0, 1)]),
        FloatPolygon([(1.2, 1.5), (0, 0.5), (1.5, 3.4)]),
    )

# This was the existing code (pretend that it's separated from the above by
# hundreds and hundreds of files):
a = old_polygon_function()

# Need to type narrow here because it could be either
# a `FloatPolygon` or an `IntPolygon`
assert isinstance(a, FloatPolygon)
a.first_float_point()

# This is the "new" code:
#
# In my case it was actually unmodified, so pretend that instead of 
# adding `new_polygon_function`, `old_polygon_function` was modified 
# to return a `BothPolygon`.

a = new_polygon_function()

# This is no longer type narrowing, because we can prove that
# `a` is a `BothPolygon`, which has no type relationship to a `FloatPolygon`.
#
# So, this assertion will always fire at runtime. It would be great to catch
# that fact at static type-checking time.
assert isinstance(a, FloatPolygon)
a.first_float_point()
```
([pyright playground](https://pyright-playground.decorator-factory.su/?gzip=H4sIAAtZ1GMC_51UsW7bMBDd9RWHZKgMMELSNouBduhQIEtRIN0MQ7pQJ4soTaokVdVAP75HSpaVxEWbaLAliu_de--Oapzdg7RakwzKGl_ggwS176wLcE8_ejKSsiyTGr2Hz9pi-Gr1YWfNOgO-amqgLJVRoSxzT7oR0Fllgl_P6E3oO02bJmIFpL_tdjXC4xVRxQiCDxM6m8kb5XwoE6pM71KVFVx9hDO8J1ZHoXdmSb653s5G7sxrbfCiAP55sQVeO29gZvxf8Z9saP-lfqymD-uF1SmkaX3Zy6dWjmg2c7x9vOFExFtODywxarG6To_MXDa9SYOVJ8fLovD7WRsm38td-Sa_Kd4KuCluVwLyawHX4x0vCHhXvF9tV1NZQ8Nfyj5LbCq0WM9ngydRXDvWS9Wmf36-4Ypi3v1SrSOSJV_Ct1Z5GNBDaAnol_JBmR2fxZog71ghmZpfYQAV3njOvUOHgWpo4pGNGHywPwkeDszV9qZ2VHtABs0PNg6fJs_9RW7U-b5EKV-IeYOFcOgIDDpnB2jJMTlJ7D2xBBbW65oXgBQXd4xCqJb2K7CO60N1SrDKeGCJPyXKK-MD8jnKUTyevQyL54f8FJAa87ng7l6kcNbZJb-7M7A_gMRRW0wRZehR80T2Zm9r1SiqBXgLj6NkFYR1TCYaqOsYeXVucioB1bnAqlTsWCGycG7TPHEgi5GqiizFfn4ulwaNBW3NjtyyAaxMzPkPxF4Ne4kdj04YXWEVwU-KChhaJVtoMfEmQkca09e9VV2U-7RxRYr03gqmjoypZ7wfBqU1oB7w4ONnjEcugOtNUHsq4I5zP87EzhG_YmqJQbYxlJh2wz2JEO58UDJpuZItye8x9UTy-vn4A5_jClG5BgAA))

**Comment by erictraut** [comment]
This example also makes use of `assert`. As explained above, conditional statements within `assert` statements are exempted from the family of "unnecessary" checks (including `reportUnnecessaryIsinstance`), and that's for good reason. We've found that they generate many false positives if enabled for asserts. It's unlikely we would change this behavior because it would generate a lot of noise in otherwise correct code.

**Comment by ottumm** [comment]
Here’s the exact same thing happening without any asserts:

```python

from collections.abc import Sequence


class FloatPolygon:
    def __init__(self, points: Sequence[tuple[float, float]]):
        self.points = points

    def first_float_point(self) -> tuple[float, float]:
        return self.points[0]


class IntPolygon:
    def __init__(self, points: Sequence[tuple[int, int]]):
        self.points = points

    def first_int_point(self) -> tuple[int, int]:
        return self.points[0]


class BothPolygon:
    def __init__(self, int_poly: IntPolygon, float_poly: FloatPolygon):
        self.int_poly = int_poly
        self.float_poly = float_poly


def old_polygon_function() -> FloatPolygon | IntPolygon:
    return FloatPolygon([(1.2, 1.5), (0, 0.5), (1.5, 3.4)])


def new_polygon_function() -> BothPolygon:
    return BothPolygon(
        IntPolygon([(0, 0), (1, 0), (0, 1)]),
        FloatPolygon([(1.2, 1.5), (0, 0.5), (1.5, 3.4)]),
    )

# This was the existing code (pretend that it's separated from the above by
# hundreds and hundreds of files):
a = old_polygon_function()

# Need to type narrow here because it could be either
# a `FloatPolygon` or an `IntPolygon`
if isinstance(a, FloatPolygon):
    a.first_float_point()

# This is the "new" code:
#
# In my case it was actually unmodified, so pretend that instead of 
# adding `new_polygon_function`, `old_polygon_function` was modified 
# to return a `BothPolygon`.

a = new_polygon_function()

# This is no longer type narrowing, because we can prove that
# `a` is a `BothPolygon`, which has no type relationship to a `FloatPolygon`.
#
# So, this assertion will always fire at runtime. It would be great to catch
# that fact at static type-checking time.
if isinstance(a, FloatPolygon):
    a.first_float_point()
```
([pyright playground](https://pyright-playground.decorator-factory.su/?gzip=H4sIAAJm1WMC_6VUwY7TMBC95ytGuwdSyRu1wF4qwYED0l4Q0nKrqsRrTxoL1w62Q6jExzN20jS7GwQLObSJ4_fmvTfj1M4eQVitUQRljS_4gwB1bK0LcI_fOjQCsywTmnsPH7Xl4bPVp4M12wzoklhDWSqjQlnmHnXNoLXKBL-d0LvQtRp3dcQySH_7_WqAxyuiigEE70Z0NpHXyvlQJlSZ3qUqK7h5Dwu8F1aHoXNmTr5b7ycjd-ZfbdAiA_p5sQVaWzYwMf6t-A82NH9SP1TTp-3M6hjSuD7v5VMrZzSZOd8-3nAhoi2XB5IYtVgt0yMxl3Vn0mDlyfG8KPx81obR93xXvss3xWsGm-J2xSBfM1gPd7TA4E3xdrVfjWUN9r8p-yyxsdBsPZ8MXkRR7VgvVRv_6XlDFdm0-6VaByRJvoYvjfLQcw-hQcAfygdlDnQWJULekkI0kl7xACq88pR7yx0PKKGORzZi-IP9jvBwIq6mM9Kh9MAJND3YOHwaPfWXU6OW-xKlfELiDRbCqUUw3DnbQ4OOyFHwziNJIGGdlrQAqKi4IxSHam6_AuuoPlSXBKtM1aC8Mj5wOkM5Z0tzx4vnp_ySkBoCuqL2XqV0ttk1vbszcDyB4IO4GCMXoeOaRrIzRytVrVAy8BYeZ0lSkMsYTXQgZcy8WhqdikG1lFiVip0rRBYKbhwoSmQ2U1WRpdyXB3Nu0FjQ1hzQzTtAytjUgB7JqyEvseXRCaErXkXwk6IM-kaJBhqeeBOhQ83T571RbZT7tHNFivTeMqKOjN6ji_uhV1oD1z0_-fgdo5kL4DoT1BELuKPcz0NxcEiviFrwIJoYSky7pp5ECLU_KJG03IgGxdeYeiL5jwH5BWjHRcu7BgAA))

**Comment by ottumm** [comment]
I neglected to edit the comment, but in my latest example the last `isinstance(a, FloatPolygon)` can never be true, no matter what type of conditional expression it’s used in.

**Comment by erictraut** [comment]
> the last `isinstance(a, FloatPolygon)` can never be true

That's not true. The type of `a` is `BothPolygon` which means that it's an instance of `BothPolygon` or a subclass thereof. It's possible to create a class that derives from both `FloatPolygon` and `BothPolygon`. You'll note that pyright narrows it type to this theoretical subclass:

```python
if isinstance(a, FloatPolygon):
    reveal_type(a) # <subclass of BothPolygon and FloatPolygon>
```

Mypy does something similar in this situation.
